kernel: Refactor process_block_header to return btck_BlockValidationState

yuvicc opened the pull request:
This PR refactors `btck_chainstate_manager_process_block_header` to return `btck_BlockValidationState` by value instead of using out-parameters or boolean returns.

hodlinator commented on src/validation.cpp:
```diff
@@ -4533,26 +4533,25 @@ bool ChainstateManager::ProcessNewBlock(const std::shared_ptr<const CBlock>& blo
 
     Chainstate* bg_chain{WITH_LOCK(cs_main, return BackgroundSyncInProgress() ? m_ibd_chainstate.get() : nullptr)};
     BlockValidationState bg_state;
     if (bg_chain && !bg_chain->ActivateBestChain(bg_state, block)) {
```
Why not use `state` here?

hodlinator replied:
But if `state` contained anything interesting, wouldn't we already have returned before reaching this block?
Callers of the function only see either `bg_state` or `state` as this PR stands anyway.

hodlinator reviewed (commented):
Concept ACK moving out-reference-parameters to return value

Not familiar with the kernel API but had a brief look at net_processing and validation.

w0xlt commented:
Concept ACK

yuvicc commented on src/validation.cpp:
```diff
@@ -4349,9 +4349,10 @@ bool ChainstateManager::AcceptBlockHeader(const CBlockHeader& block, BlockValida
+BlockValidationState ChainstateManager::ProcessNewBlockHeaders(std::span<const CBlockHeader> headers, bool min_pow_checked, const CBlockIndex** ppindex)
 {
     AssertLockNotHeld(cs_main);
+    BlockValidationState state;
```
Correct, we will definitely need that.

yuvicc commented:
Thanks for the review @hodlinator 
- Added `Assume(!headers.empty())` before instantiating validation state as suggested in [comment](https://github.com/bitcoin/bitcoin/pull/33856#discussion_r2517354777)
- Removed shadow `state` [comment](https://github.com/bitcoin/bitcoin/pull/33856#discussion_r2517333716)

yancyribbens commented on src/test/blockfilter_index_tests.cpp:
```diff
@@ -104,8 +104,7 @@ bool BuildChainTestingSetup::BuildChain(const CBlockIndex* pindex,
 
-        BlockValidationState state;
-        if (!Assert(m_node.chainman)->ProcessNewBlockHeaders({{header}}, true, state, &pindex)) {
+        if (BlockValidationState state{Assert(m_node.chainman)->ProcessNewBlockHeaders({{header}}, true, &pindex)}; !state.IsValid()) {
```
in 1f589e1af162c2c2705f0404496c549785f2f545 could `!state.IsValud()` be `state.IsInvalid()`?

w0xlt reviewed (commented):
This PR eliminates the ambiguity of the boolean return value and improves the clarity of both the internal and kernel APIs.

ACK https://github.com/bitcoin/bitcoin/pull/33856/commits/1f589e1af162c2c2705f0404496c549785f2f545 (with the above-mentioned nit)

yuvicc commented:
Thanks for the review @w0xlt 
- Addressed the [suggestion](https://github.com/bitcoin/bitcoin/pull/33856#discussion_r2547248797) and added as co-author

TheCharlatan commented on src/validation.cpp:
```diff
@@ -4505,14 +4507,15 @@ bool ChainstateManager::AcceptBlock(const std::shared_ptr<const CBlock>& pblock,
 }
 
-bool ChainstateManager::ProcessNewBlock(const std::shared_ptr<const CBlock>& block, bool force_processing, bool min_pow_checked, bool* new_block)
+BlockValidationState ChainstateManager::ProcessNewBlock(const std::shared_ptr<const CBlock>& block, bool force_processing, bool min_pow_checked, bool* new_block)
```
I don't think this change is correct. Previously false was returned when the block failed to be accepted or on a fatal error condition, not when it failed to validate. If you look into `ActivateBestChainStep`, you'll see that we break on a validation failure, reset the state again and don't return false.  Can you drop this change again?

sedited commented on src/validation.cpp:
```diff
@@ -4542,17 +4550,23 @@ bool ChainstateManager::ProcessNewBlock(const std::shared_ptr<const CBlock>& blo
 
-    return true;
+    // Attempts to connect the block to the tip of the active chain.
+    BlockValidationState activation_state;
```
I'm confused by this change. Why are you introducing this variable? The comment is also wrong, since this processes the block on the background chain. I would still prefer if the behaviour of ProcessNewBlock would not be changed as part of this patch set.

sedited replied:
The comment you linked seems misleading. `bg_state` is the correct name in my view. We are not operating on the active chain here.

I think the belt-and-suspenders check in `ActivateBestChain` is broken by this change. When we return false in the case of the chain being disabled, we now run into `NONFATAL_UNREACHABLE`. It might be true that there are no similar cases in the call graph of `ActivateBestChain`, but can we guarantee that? For this reason, I think the commit should be dropped.

danielabrozzoni reviewed (commented):
light ACK f31f7c21ff7093a4c90199cd0bb2128d19bf2d33

Code looks good to me, and the interface is cleaner now. I'm only light-acking sicne I'm not familiar with the kernel :)

As I pointed out in a comment, I don't think we needed the `Assume(!headers.empty())` at the start of `ProcessNewBlockHeaders`, but I'm ok with keeping it.

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/33856/commits/be379fd52b82805ab99400457ea44f93430f16e2

stickies-v reviewed (commented):
Concept ACK. As commented, I think it makes sense to use `util::Expected` here.

yuvicc commented on src/validation.h:
```diff
@@ -1239,11 +1239,12 @@ class ChainstateManager
+     *          validation failure reason.
      */
-    bool ProcessNewBlockHeaders(std::span<const CBlockHeader> headers, bool min_pow_checked, BlockValidationState& state, const CBlockIndex** ppindex = nullptr) LOCKS_EXCLUDED(cs_main);
+    BlockValidationState ProcessNewBlockHeaders(std::span<const CBlockHeader> headers, bool min_pow_checked, const CBlockIndex** ppindex = nullptr) LOCKS_EXCLUDED(cs_main);
```
Maybe we can also return state as a param for callers who need any failure reason and returning `btck_BlockTreeEntry` as the natural return for chaining. I am not sure here. 
>I think btck_chainstate_manager_process_block_header returning a btck_BlockTreeEntry* (null if header could not be added to the block tree) makes for a much more natural and chainable API.
>

stringintech replied:
> I'm not sure how enthusiastic I am about `BlockValidationState` representing both validation states and runtime errors, but I'll need to look into it more.

Yeah... I'm not sure why runtime errors ended up mixed into `ValidationState` in the first place. I experimented with removing `M_ERROR` mode entirely here: https://github.com/stringintech/bitcoin/commit/beaa25ff76626fddf54f23f63adbc8a87ac2417c. 
Functions like `FlushStateToDisk`, `ActivateBestChain` aren't exactly about validating a certain block, yet they take a `BlockValidationState&` out-param just to absorb runtime errors, which feels like the wrong abstraction. Since we seem to prefer error handling explicit at call sites, `util::Expected` seems like a natural fit (over throwing exceptions). Those functions could return `util::Expected<void, std::string>` and drop the out-param entirely, while functions that can both perform validation and hit a runtime error (like `ConnectBlock`, `AcceptBlock`) could return `util::Expected<BlockValidationState, std::string>`, making the two failure modes distinct. The changes feel fairly mechanical, though I'm not sure it's worth the review churn as a standalone effort.

hodlinator reviewed (approved):
ACK 697b8a78028322a91738da47bc06cec0ddd544a9

This PR basically removes the second column for `btck_chainstate_manager_process_block_header`:

| state | int return value |
|:---|---:|
| M_VALID | 0 |
| M_INVALID | -1 |
| M_ERROR | -1 |
| nullptr (exception) | -1 |

The `int` return value was kind of nice to have for the happy path, the PR requires consumers of the API to check for `nullptr` state before checking for M_VALID. Good for rigor, slight friction for prototyping. One less out-parameter is nice.

fanquake commented on src/test/validation_block_tests.cpp:
```diff
@@ -3,14 +3,14 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <boost/test/unit_test.hpp>
+#include <boost/test/unit_test_suite.hpp>
```
IWYU isn't enforced on `src/test` yet, but it's saying to remove this line in any case:
https://github.com/bitcoin/bitcoin/actions/runs/23594588120/job/68707982908?pr=33856#step:11:28669.

yuvicc replied:
Done. Thanks.

stringintech commented on src/kernel/bitcoinkernel.cpp:
```diff
@@ -1308,19 +1298,18 @@ int btck_chainstate_manager_process_block(
     try {
         auto& chainman = btck_ChainstateManager::get(chainstate_manager).m_chainman;
-        auto result = chainman->ProcessNewBlockHeaders({&btck_BlockHeader::get(header), 1}, /*min_pow_checked=*/true, btck_BlockValidationState::get(state), /*ppindex=*/nullptr);
+        auto state = chainman->ProcessNewBlockHeaders({&btck_BlockHeader::get(header), 1}, /*min_pow_checked=*/true, /*ppindex=*/nullptr);
```
nit
```suggestion
        auto state = chainman->ProcessNewBlockHeaders({&btck_BlockHeader::get(header), 1}, /*min_pow_checked=*/true);
```

stringintech commented on src/test/validation_block_tests.cpp:
```diff
@@ -3,14 +3,14 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <boost/test/unit_test.hpp>
+#include <boost/test/unit_test_suite.hpp>
```
Seems this line caused the lint job to fail (as already mentioned in https://github.com/bitcoin/bitcoin/pull/33856#discussion_r2994842186).

stringintech commented on src/net_processing.cpp:
```diff
@@ -3076,35 +3050,33 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
-    // Now process all the headers.
-    BlockValidationState state;
-    const bool processed{m_chainman.ProcessNewBlockHeaders(headers,
+    // Now process all the headers and return the block validation state.
```
nit: added comment seems redundant.
```suggestion
    // Now process all the headers.
```

stringintech commented on src/net_processing.cpp:
```diff
@@ -3076,35 +3050,33 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
-    BlockValidationState state;
-    const bool processed{m_chainman.ProcessNewBlockHeaders(headers,
+    // Now process all the headers and return the block validation state.
+    BlockValidationState state{m_chainman.ProcessNewBlockHeaders(headers,
```
Although `ProcessNewBlockHeaders` doesn't seem to set state to `M_ERROR` mode, I think it is better to not rely on that assumption - similar to how `submitheader` explicitly handles the error case when calling the same function. Perhaps we can have sth like:

<details>
<summary>diff</summary>

```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index a0273f7d82..da9ff706f3 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -3054,15 +3054,18 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
     BlockValidationState state{m_chainman.ProcessNewBlockHeaders(headers,
                                                            /*min_pow_checked=*/true,
                                                            &pindexLast)};
-    if (state.IsInvalid()) {
-        if (!pfrom.IsInboundConn() && state.GetResult() == BlockValidationResult::BLOCK_CACHED_INVALID) {
-            // Warn user if outgoing peers send us headers of blocks that we previously marked as invalid.
-            LogWarning("%s (received from peer=%i). "
-                        "If this happens with all peers, consider database corruption (that -reindex may fix) "
-                        "or a potential consensus incompatibility.",
-                        state.GetDebugMessage(), pfrom.GetId());
-        }
-        MaybePunishNodeForBlock(pfrom.GetId(), state, via_compact_block, "invalid header received");
+    if (!state.IsValid()) {
+        // ProcessNewBlockHeaders only sets Invalid, never Error state.
+        if (Assume(state.IsInvalid())) {
+            if (!pfrom.IsInboundConn() && state.GetResult() == BlockValidationResult::BLOCK_CACHED_INVALID) {
+                // Warn user if outgoing peers send us headers of blocks that we previously marked as invalid.
+                LogWarning("%s (received from peer=%i). "
+                            "If this happens with all peers, consider database corruption (that -reindex may fix) "
+                            "or a potential consensus incompatibility.",
+                            state.GetDebugMessage(), pfrom.GetId());
+            }
+            MaybePunishNodeForBlock(pfrom.GetId(), state, via_compact_block, "invalid header received");
+        }
         return;
     }
     assert(pindexLast);
@@ -4560,8 +4563,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
         const CBlockIndex *pindex = nullptr;
         BlockValidationState state{m_chainman.ProcessNewBlockHeaders({{cmpctblock.header}}, /*min_pow_checked=*/true, &pindex)};
-        if (state.IsInvalid()) {
-            MaybePunishNodeForBlock(pfrom.GetId(), state, /*via_compact_block=*/true, "invalid header via cmpctblock");
+        if (!state.IsValid()) {
+            // ProcessNewBlockHeaders only sets Invalid, never Error state.
+            if (Assume(state.IsInvalid())) {
+                MaybePunishNodeForBlock(pfrom.GetId(), state, /*via_compact_block=*/true, "invalid header via cmpctblock");
+            }
             return;
         }
 

```
</details>

yuvicc replied:
Agree!

stringintech commented on src/net_processing.cpp:
```diff
@@ -3076,35 +3050,33 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
     assert(pindexLast);
 
-    if (processed && received_new_header) {
+    if (state.IsValid() && received_new_header) {
```
In combination with above comment, the first check seems redundant:
```suggestion
    if (received_new_header) {
```
A related discussion is here https://github.com/bitcoin/bitcoin/pull/27826#discussion_r1235685769.

yuvicc commented:
Thanks for review @stringintech 
- Fixed nits: https://github.com/bitcoin/bitcoin/pull/33856#discussion_r3074190956, https://github.com/bitcoin/bitcoin/pull/33856#discussion_r3073991538, https://github.com/bitcoin/bitcoin/pull/33856#discussion_r3073970428 & https://github.com/bitcoin/bitcoin/pull/33856#discussion_r3073942210
- Explicitly handle the error case for state. [comment](https://github.com/bitcoin/bitcoin/pull/33856#discussion_r3074186363)

yuvicc commented on src/test/validation_block_tests.cpp:
```diff
@@ -368,4 +367,9 @@ BOOST_AUTO_TEST_CASE(witness_commitment_index)
     BOOST_CHECK_EQUAL(GetWitnessCommitmentIndex(pblock), 2);
 }
+
+BOOST_AUTO_TEST_CASE(test_empty_process_new_block_headers) {
```
Done. Thanks.

w0xlt reviewed (commented):
reACK 93ea1b76686552b16486e55dc4e081f4b2661723

optout21 commented on src/net_processing.cpp:
```diff
@@ -4485,7 +4547,7 @@ void PeerManagerImpl::ProcessMessage(Peer& peer, CNode& pfrom, const std::string
         if (!prev_block) {
             // Doesn't connect (or is genesis), instead of DoSing in AcceptBlockHeader, request deeper headers
```
> [93ea1b7](https://github.com/bitcoin/bitcoin/pull/33856/changes/93ea1b76686552b16486e55dc4e081f4b2661723) _validation: Return BlockValidationState from ProcessNewBlockHeaders_:

A nit outside the scope of this PR: "AcceptBlockHeader" in the comment (L4548) seems to be outdated and seems to refer to the `ProcessNewBlockHeaders` call below (L4565). Similarly, the comment on L4574.

Consider if it makes sense to check these in this PR, and if so, consider if they are indeed stale.

optout21 reviewed (commented):
ACK 93ea1b76686552b16486e55dc4e081f4b2661723

A refactor of `ProcessNewBlockHeaders()` parameters, to return validation state in return value instead of an output parameter. The redundant return value is removed. The kernel wrapper is also updated.
The change set is compact, self-sufficient and well-rounded. I have not found any issue with it.

What I did: Review the code; check compile and unit tests locally. Check the call sites, and their usage of the result. Check `BlockValidationState` and its usage in general.

Left a non-blocking nit regarding potentially stale related comments.

A note with scope beyond this PR: there are several other methods that return state in a similar fashion (in an output parameter), such as `AcceptBlockHeader()`.

stringintech commented:
ACK 93ea1b76

I’m a bit uneasy that the C API can return `nullptr` for runtime errors while `BlockValidationState` could also carry error state. Feels like two failure channels in one abstraction. Would be nice to split validation result from runtime failure more cleanly in the future, maybe along the lines of https://github.com/bitcoin/bitcoin/pull/33856#discussion_r3068580118.

w0xlt reviewed (commented):
reACK 93ea1b76686552b16486e55dc4e081f4b2661723

hodlinator commented on src/net_processing.cpp:
```diff
@@ -3078,34 +3051,35 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
-                           state.GetDebugMessage(), pfrom.GetId());
+                            "If this happens with all peers, consider database corruption (that -reindex may fix) "
+                            "or a potential consensus incompatibility.",
+                            state.GetDebugMessage(), pfrom.GetId());
```
Adding this new 1-space indentation for these 3 lines makes the code worse and messes with git blame history. No other reviewer caught this, and not the author? Oh, I see I also didn't catch it in my A-C-K of the earlier 697b8a78028322a91738da47bc06cec0ddd544a9 which also has this issue.

I think it's fair to invalidate A-C-Ks to undo this.

hodlinator commented on src/net_processing.cpp:
nit: Could align indentation:
```diff
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -3052,8 +3052,8 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
 
     // Now process all the headers.
     BlockValidationState state{m_chainman.ProcessNewBlockHeaders(headers,
-                                                           /*min_pow_checked=*/true,
-                                                           &pindexLast)};
+                                                                 /*min_pow_checked=*/true,
+                                                                 &pindexLast)};
     if (!state.IsValid()) {
         // ProcessNewBlockHeaders only sets Invalid, never Error state.
         if (Assume(state.IsInvalid())) {
```

hodlinator reviewed (commented):
Approach NACK 93ea1b76686552b16486e55dc4e081f4b2661723

I regret A-C-King this change now, it feels incomplete. Sorry for changing my mind.

Adding all the checks that we're either Valid or Invalid but not in Error may be technically correct in the present state but feels like laying land mines. Even if it would grow the PR a lot I would rather see a refactoring which makes the Error state impossible, along with `AcceptBlock()` et all not both returning a `bool` and a state which could theoretically mismatch. As it is we are making the validation code harder to maintain in order to please the kernel API.

hodlinator commented:
Started going down the rabbit hole of implementing something more fully baked and making unexpected states impossible instead of adding a bunch of checks. Didn't find a really elegant way in my quick exploration but there may be one.

I would prefer this change was cut back to only touching the kernel API, or do a deeper change as I suggested above.

<details><summary>Suggested minimal diff that I would feel more comfortable A-C-King</summary>

```diff
diff --git a/src/kernel/bitcoinkernel.cpp b/src/kernel/bitcoinkernel.cpp
index ea646cd551..50ed46cd0c 100644
--- a/src/kernel/bitcoinkernel.cpp
+++ b/src/kernel/bitcoinkernel.cpp
@@ -1298,19 +1298,19 @@ int btck_chainstate_manager_process_block(
     return result ? 0 : -1;
 }
 
-int btck_chainstate_manager_process_block_header(
+btck_BlockValidationState* btck_chainstate_manager_process_block_header(
     btck_ChainstateManager* chainstate_manager,
-    const btck_BlockHeader* header,
-    btck_BlockValidationState* state)
+    const btck_BlockHeader* header)
 {
     try {
         auto& chainman = btck_ChainstateManager::get(chainstate_manager).m_chainman;
-        auto result = chainman->ProcessNewBlockHeaders({&btck_BlockHeader::get(header), 1}, /*min_pow_checked=*/true, btck_BlockValidationState::get(state), /*ppindex=*/nullptr);
-
-        return result ? 0 : -1;
+        auto state = btck_BlockValidationState::create();
+        auto result = chainman->ProcessNewBlockHeaders({&btck_BlockHeader::get(header), 1}, /*min_pow_checked=*/true, btck_BlockValidationState::get(state));
+        assert(result == btck_BlockValidationState::get(state).IsValid());
+        return state;
     } catch (const std::exception& e) {
         LogError("Failed to process block header: %s", e.what());
-        return -1;
+        return nullptr;
     }
 }
 
diff --git a/src/kernel/bitcoinkernel.h b/src/kernel/bitcoinkernel.h
index 5427e77617..f6eb4c3b70 100644
--- a/src/kernel/bitcoinkernel.h
+++ b/src/kernel/bitcoinkernel.h
@@ -1117,13 +1117,11 @@ BITCOINKERNEL_API const btck_BlockTreeEntry* BITCOINKERNEL_WARN_UNUSED_RESULT bt
  *
  * @param[in] chainstate_manager        Non-null.
  * @param[in] header                    Non-null btck_BlockHeader to be validated.
- * @param[out] block_validation_state   The result of the btck_BlockHeader validation.
- * @return                              0 if btck_BlockHeader processing completed successfully, non-zero on error.
+ * @return                              The btck_BlockValidationState containing validation results, or null on error.
  */
-BITCOINKERNEL_API int BITCOINKERNEL_WARN_UNUSED_RESULT btck_chainstate_manager_process_block_header(
+BITCOINKERNEL_API btck_BlockValidationState* BITCOINKERNEL_WARN_UNUSED_RESULT btck_chainstate_manager_process_block_header(
     btck_ChainstateManager* chainstate_manager,
-    const btck_BlockHeader* header,
-    btck_BlockValidationState* block_validation_state) BITCOINKERNEL_ARG_NONNULL(1, 2, 3);
+    const btck_BlockHeader* header) BITCOINKERNEL_ARG_NONNULL(1, 2);
 
 /**
  * @brief Triggers the start of a reindex if the wipe options were previously
diff --git a/src/kernel/bitcoinkernel_wrapper.h b/src/kernel/bitcoinkernel_wrapper.h
index 064d0dd14b..632a89d0d2 100644
--- a/src/kernel/bitcoinkernel_wrapper.h
+++ b/src/kernel/bitcoinkernel_wrapper.h
@@ -940,6 +940,8 @@ public:
     explicit BlockValidationState() : Handle{btck_block_validation_state_create()} {}
 
     BlockValidationState(const BlockValidationStateView& view) : Handle{view} {}
+
+    BlockValidationState(btck_BlockValidationState* state) : Handle{state} {}
 };
 
 class ValidationInterface
@@ -1217,9 +1219,13 @@ public:
         return res == 0;
     }
 
-    bool ProcessBlockHeader(const BlockHeader& header, BlockValidationState& state)
+    BlockValidationState ProcessBlockHeader(const BlockHeader& header)
     {
-        return btck_chainstate_manager_process_block_header(get(), header.get(), state.get()) == 0;
+        auto state = btck_chainstate_manager_process_block_header(get(), header.get());
+        if (!state) {
+            throw std::runtime_error("Failed to process block header");
+        }
+        return BlockValidationState{state};
     }
 
     ChainView GetChain() const
diff --git a/src/test/kernel/test_kernel.cpp b/src/test/kernel/test_kernel.cpp
index 5a38006539..af29942058 100644
--- a/src/test/kernel/test_kernel.cpp
+++ b/src/test/kernel/test_kernel.cpp
@@ -1013,10 +1013,9 @@ BOOST_AUTO_TEST_CASE(btck_chainman_regtest_tests)
         for (const auto& data : REGTEST_BLOCK_DATA) {
             Block block{hex_string_to_byte_vec(data)};
             BlockHeader header = block.GetHeader();
-            BlockValidationState state{};
-            BOOST_CHECK(state.GetBlockValidationResult() == BlockValidationResult::UNSET);
-            BOOST_CHECK(chainman->ProcessBlockHeader(header, state));
+            BlockValidationState state = chainman->ProcessBlockHeader(header);
             BOOST_CHECK(state.GetValidationMode() == ValidationMode::VALID);
+            BOOST_CHECK(state.GetBlockValidationResult() == BlockValidationResult::UNSET);
             BlockTreeEntry entry{*chainman->GetBlockTreeEntry(header.Hash())};
             BOOST_CHECK(!chainman->GetChain().Contains(entry));
             BlockTreeEntry best_entry{chainman->GetBestEntry()};
```

</details>

Apologies again for raising my bar after you already passed over it.

yuvicc commented:
> 
> Adding all the checks that we're either Valid or Invalid but not in Error may be technically correct in the present state but feels like laying land mines. Even if it would grow the PR a lot I would rather see a refactoring which makes the Error state impossible, along with `AcceptBlock()` et all not both returning a `bool` and a state which could theoretically mismatch.

I guess I agree with you @hodlinator and @stringintech, for this PR specifically I would like to keep it to only kernel API rather than refactoring the whole validation code which could be done in another PR which would also reduce the headache while reviewing the PR and separation of concerns.

yuvicc commented:
Thanks for the review @hodlinator I have updated the PR title and description as well as the suggestions you mentioned above and added you as co-author.

w0xlt reviewed (commented):
reACK 2471404f7d17898e1f5792b7db1bc2b10f837b47

hodlinator commented on src/kernel/bitcoinkernel.cpp:
nit: Commit message has typo "ProcessNewBlockHeadersalready\nalready".

hodlinator commented on src/kernel/bitcoinkernel.cpp:
```diff
@@ -1340,19 +1298,20 @@ int btck_chainstate_manager_process_block(
 
-        return result ? 0 : -1;
+        auto state = btck_BlockValidationState::create();
+        auto result = chainman->ProcessNewBlockHeaders({&btck_BlockHeader::get(header), 1}, /*min_pow_checked=*/true, btck_BlockValidationState::get(state));
```
nanonit: Could make `bool` explicit to reader and use curly-braces to prevent narrowing:
```suggestion
        bool result{chainman->ProcessNewBlockHeaders({&btck_BlockHeader::get(header), 1}, /*min_pow_checked=*/true, btck_BlockValidationState::get(state))};
```

stickies-v replied:
+1

hodlinator reviewed (approved):
ACK 2471404f7d17898e1f5792b7db1bc2b10f837b47

Thanks for taking my feedback @yuvicc.

I stared some more today at your previous version of the PR, 93ea1b76686552b16486e55dc4e081f4b2661723 - specifically the changes to net_processing.cpp and validation.cpp. There is some nuance to them that probably make them less scary than I was previously thinking. I was somehow hallucinating:

```C++
if (!state.IsValid()) {
    CHECK_NONFATAL(state.IsInvalid());
```
or
```C++
if (!accepted) {
    if (!state.IsInvalid()) NONFATAL_UNREACHABLE();
```
But your change at least allowed for `BlockValidationState::M_ERROR` in the upper part of `ChainstateManager::ProcessNewBlockHeaders()`.

On the other hand you added `Assume()`s in net_processing.cpp which some seem to have issues with in other contexts: https://github.com/bitcoin/bitcoin/pull/34440#discussion_r3085393412

---

Was coming to terms with 93ea1b76686552b16486e55dc4e081f4b2661723 minus the whitespace issues and the `Assume()`s. I then had a health issue and was unable to finish my re-review earlier today.

Here is my experimental branch which builds upon your earlier push and consistently returns `BlockValidationState` or `NoErrorBlockValidationState` from functions: https://github.com/bitcoin/bitcoin/compare/master...hodlinator:bitcoin:pr/33856_suggestions
It's a bit of a beast to polish & review, and should probably be broken up into many commits.

That experiment makes your inclusion of the refactor of `ChainstateManager::ProcessNewBlockHeaders()` look like a valid strategic move. Sorry for the onset of paranoia.

yuvicc commented:
Thanks @hodlinator I will take a look at this and push when ready.
> Here is my experimental branch which builds upon your earlier push and consistently returns BlockValidationState or NoErrorBlockValidationState from functions: https://github.com/bitcoin/bitcoin/compare/master...hodlinator:bitcoin:pr/33856_suggestions
It's a bit of a beast to polish & review, and should probably be broken up into many commits.

optout21 commented:
ACK 88d9bc5aa4f9e054487e8cfa60a1f12c5d508237
Re-acked, only minor changes since my last review.

Prev:
ACK 2471404f7d17898e1f5792b7db1bc2b10f837b47

I've processed the new discussion since my [last review](https://github.com/bitcoin/bitcoin/pull/33856#pullrequestreview-4119044329), re-reviewed the changes. The change-set has been narrowed down, restricted to kernel part. LGTM in the current form. However, from the discussions I gather that it may still change in scope (that's fine).

optout21 commented:
> Here is my experimental branch which builds upon your earlier push and consistently returns `BlockValidationState` or `NoErrorBlockValidationState` from functions

@hodlinator (cc @yuvicc) : I've checked your branch, I've started something very similar 1-2 weeks ago, visible here:
https://github.com/optout21/bitcoin/pull/7
I've started changing the validation state output parameter to a return value in a few methods. I meant it as a continuation of this PR, but in fact it can go independently. I may improve/finalize it and open into a new PR, unless there are objections.

stickies-v commented on src/kernel/bitcoinkernel_wrapper.h:
```diff
@@ -1006,37 +940,10 @@ class BlockValidationState : public Handle<btck_BlockValidationState, btck_block
-    {
-        return static_cast<TxValidationResult>(btck_tx_validation_state_get_tx_validation_result(get()));
-    }
+    BlockValidationState(btck_BlockValidationState* state) : Handle{state} {}
```
nit: I think this ctor (and above) should be `explicit`

stickies-v commented on src/kernel/bitcoinkernel_wrapper.h:
```diff
@@ -1317,9 +1219,13 @@ class ChainMan : UniqueHandle<btck_ChainstateManager, btck_chainstate_manager_de
-        return btck_chainstate_manager_process_block_header(get(), header.get(), state.get()) == 0;
+        auto state = btck_chainstate_manager_process_block_header(get(), header.get());
+        if (!state) {
+            throw std::runtime_error("Failed to process block header");
```
I don't think we need this, `Handle` [already calls](https://github.com/stickies-v/bitcoin/blob/7844a2f08303edb665e4292d002a826f24e62832/src/kernel/bitcoinkernel_wrapper.h#L168) `check` in its ctor.

hodlinator replied:
Since I'm unfamiliar with this code I confirmed this by always setting the pointer to `nullptr` and running test_kernel and got the expected exception. This appears to be the call:
https://github.com/bitcoin/bitcoin/blob/88d9bc5aa4f9e054487e8cfa60a1f12c5d508237/src/kernel/bitcoinkernel_wrapper.h#L315

stickies-v reviewed (commented):
code LGTM 2471404f7d17898e1f5792b7db1bc2b10f837b47 modulo outstanding comments. I think this is a clear, contained interface win and think we can make other changes in follow-ups. Would prefer to undraft this and merge ~as-is (after addressing outstanding comments)

hodlinator commented:
Agree with @stickies-v that the best way forward for this PR would be to as-is (+ potentially addressing some nits), maybe also bringing back some of the changes from 93ea1b76686552b16486e55dc4e081f4b2661723 which I was paranoid about but was coming to terms with.

I would caution against doing the full refactor to return `BlockValidationState` from `AcceptBlockHeader()` etc as me and @optout21 experimented with as part of this PR. Happy to help those changes along in another PR, whoever authors it.

yuvicc committed:
kernel: Return btck_BlockValidationState from process_block_header API

Remove redundant int return from btck_chainstate_manager_process_block_header.
Previously returned both an int result and an output validation state parameter, creating ambiguity
where non-zero could mean either invalid header or processing failure. Since ProcessNewBlockHeaders
already provides complete validation info, the int return was redundant.

Co-authored-by: stringintech <stringintech@gmail.com>
Co-authored-by: stickies-v <stickies-v@protonmail.com>
Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

yuvicc commented:
I have addressed all the nits and marked this as ready for review. Further changes can be done in a follow-up PR. Thanks @stickies-v and @hodlinator.

stickies-v reviewed (approved):
ACK 88d9bc5aa4f9e054487e8cfa60a1f12c5d508237

PR description needs updating: `return btck_BlockValidationState by value ` - we return by ptr.

w0xlt reviewed (commented):
reACK 88d9bc5aa4f9e054487e8cfa60a1f12c5d508237

hodlinator reviewed (approved):
re-ACK 88d9bc5aa4f9e054487e8cfa60a1f12c5d508237

Only addressed nits since previous review.

optout21 commented:
> Happy to help those changes along in another PR, whoever authors it.

A followup PR #35570 has been opened. @hodlinator , I'd appreciate your review.
